### PR TITLE
Map fusion

### PR DESCRIPTION
**Pull request by nikivazou**
Proving map fusion

### CODE REVIEW

File: src/Language/Haskell/Liquid/Constraint/Generate.hs [hunk 1]
```diff
@@ -1551,6 +1551,7 @@ funExpr _ _
 
 untick :: CoreExpr -> CoreExpr
 untick (Tick _ e)  = untick e 
+untick (App e (Type _)) = untick e
```
Comment (gridaphobe) [inline src/Language/Haskell/Liquid/Constraint/Generate.hs]: This looks bogus, why are you removing the `Type` argument?

Comment (nikivazou) [inline src/Language/Haskell/Liquid/Constraint/Generate.hs]: Let me rename: I want to get 

compose @a @b @c f g :: {v:_ | v == compose f g}

**Comment by nikivazou** [comment]
Any idea how one can write the natural map function and get the following logic-friendly version?

```
map f xs 
  | nill xs   = N 
  | otherwise = C (f (hd xs)) (map f (tl xs)) 
```

**Comment by nikivazou** [comment]
@ranjitjhala here are the summaries:

[develop.csv](https://github.com/ucsd-progsys/liquidhaskell/blob/08d9fa6923e66df9d1e01dcece0fb05875c0f876/tests/log/summary-develop.csv) VS [map_fusion.csv](https://github.com/ucsd-progsys/liquidhaskell/blob/08d9fa6923e66df9d1e01dcece0fb05875c0f876/tests/log/summary-map_fusion.csv)

There is a small slowdown, I assume because I am convertion logic in a curry form. Can I merge?

**Comment by ranjitjhala** [comment]
Can't the conversion be guarded by the allowHO flag?

**Comment by nikivazou** [comment]
It could, but better if the whole defunctionalization is guarded by the allowHO flag. 

Since the slowdown is not much, can I merge now and continue with the monadic laws and guard the defunctionalization step later?

**Comment by ranjitjhala** [comment]
OK let me look at the tables and get back to you in like 20 minutes

**Pull request merged**